submarine-vis/src/components/App.js

```diff
@@ -16,11 +17,42 @@ class App extends Component {
       isSafetyConditionAchieved: true,
       submarineRegion: 'yellow',
       trenchAlert: 'red',
-      probes: [96, 4, 7],
+      probes: [],
+      gamestates: [],
+      interval: undefined,
     };
   }
 
-  // TODO: have a function that talks to the server for game stats
+  componentDidMount() {
+    this.setState({
+        interval: setInterval(() => {
+          const gamestate = this.state.gamestates.shift()
+          if (gamestate !== undefined) {
+            this.setState({
+              redRegion: gamestate.red_region,
+              position: gamestate.position,
+              isSafetyConditionAchieved: gamestate.is_safety_condition_achieved,
+              submarineRegion: gamestate.submarine_region,
+              trenchAlert: gamestate.trench_alert,
+              probes: gamestate.probes,
+            })
+          }
+        }, 1000)
+    })
+  }
+
+  componentWillUnmount() {
+    if (this.state.interval !== undefined) {
+      clearInterval(this.state.interval)
```

Comment:
I'm pretty sure you can mutate a state without using setState. It can just cause problems. This isn't mutating the state though. It's taking the interval id from the state and stopping the async interval.